src/os_handlers.eliom

```diff
@@ -202,63 +205,90 @@ let%client connect_handler_rpc =
        connect_handler_rpc')
 let%client connect_handler () v = connect_handler_rpc v
 
-let activation_handler_common ~akey =
+[%%shared
+  exception Custom_action_link of
+      Os_data.actionlinkkey_info
+      * bool (* If true, the link corresponds to a phantom user
+                (user who never created its account).
+                In that case, you probably want to display a sign-up form,
+                and in the other case a login form. *)
+]
+
+let action_link_handler_common akey =
   (* <s>
      SECURITY: we disconnect the user before doing anything.
      If the user is already connected,
-     we're going to disconnect him even if the activation key outdated.</s>
+     we're going to disconnect him even if the action_link key outdated.</s>
      ---> Now we disconnect the user only if we reconnect them because it's
      only annoying to disconnect people with unvalid keys.
-     TODO: do not disconnect users if we relog them with the same userid.
   *)
+  let myid_o = Os_current_user.Opt.get_current_userid () in
   try%lwt
-    let%lwt {Os_user.userid; email; validity; action; data = _; autoconnect} =
-      Os_user.get_activationkey_info akey in
+    let%lwt
+      {Os_data.userid; email; validity; action; data = _; autoconnect}
+      as action_link =
+      Os_user.get_actionlinkkey_info akey
+    in
     let%lwt () =
-      if action = `AccountActivation && validity <= 0L then
-        Lwt.fail Os_db.Account_already_activated
-      else Lwt.return_unit in
+      if action = `AccountActivation && validity <= 0L
+      then Lwt.fail Os_db.Account_already_activated
+      else Lwt.return_unit
+    in
     let%lwt () =
-      if validity <= 0L then
-        Lwt.fail Os_db.No_such_resource
-      else Lwt.return_unit in
-    let%lwt () = Os_db.User.set_email_validated userid email in
-    if autoconnect then
+      if validity <= 0L
+      then Lwt.fail Os_db.No_such_resource
+      else Lwt.return_unit
+    in
+    let%lwt () =
+      if action = `AccountActivation
+      then Os_db.User.set_email_validated userid email
+      else Lwt.return_unit
+    in
+    if autoconnect && myid_o <> Some userid
+    then
       let%lwt () = Os_session.disconnect () in
       let%lwt () = Os_session.connect userid in
       Lwt.return `Restart
     else
-      Lwt.return `Reload
+      match action with
```

Comment:
Well, I'm OK with leaving this here, as long as it's a clear choice and we document it.